src/renderer/dx/DxSoftFont.cpp

```diff
@@ -0,0 +1,256 @@
+constexpr size_t BITMAP_GRID_HEIGHT = 8;
+constexpr size_t PADDING = 2;
+
+constexpr auto ANTIALIASED_INTERPOLATION = D2D1_SCALE_INTERPOLATION_MODE_HIGH_QUALITY_CUBIC;
+constexpr auto ALIASED_INTERPOLATION = D2D1_SCALE_INTERPOLATION_MODE_NEAREST_NEIGHBOR;
+
+DxSoftFont::DxSoftFont() noexcept :
+    _centeringHint{},
+    _interpolation{ ANTIALIASED_INTERPOLATION },
+    _colorMatrix{}
+{
+    _colorMatrix.m[0][3] = 1;
+}
+
+void DxSoftFont::SetFont(const gsl::span<const uint16_t> bitPattern,
+                         const til::size sourceSize,
+                         const til::size targetSize,
+                         const size_t centeringHint)
+{
+    Reset();
+
+    // If the font is being reset, just free up the memory and return.
+    if (bitPattern.empty())
+    {
+        _bitmapBits.clear();
+        return;
+    }
+
+    const auto maxGlyphCount = BITMAP_GRID_WIDTH * BITMAP_GRID_HEIGHT;
+    _glyphCount = std::min<size_t>(bitPattern.size() / sourceSize.height, maxGlyphCount);
+    _sourceSize = sourceSize;
+    _targetSize = targetSize;
+    _centeringHint = centeringHint;
+
+    const auto bitmapWidth = BITMAP_GRID_WIDTH * (_sourceSize.width + PADDING * 2);
+    const auto bitmapHeight = BITMAP_GRID_HEIGHT * (_sourceSize.height + PADDING * 2);
+    _bitmapBits = std::vector<UINT32>(bitmapWidth * bitmapHeight);
+    _bitmapSize = { gsl::narrow_cast<UINT32>(bitmapWidth), gsl::narrow_cast<UINT32>(bitmapHeight) };
+
+    const auto bitmapScanline = [=](const auto lineNumber) noexcept {
+        return _bitmapBits.begin() + lineNumber * bitmapWidth;
+    };
+
+    // The source bitPattern is just a list of the scanlines making up the
+    // glyphs one after the other, but we want to lay them out in a grid, so
+    // we need to process each glyph individually.
+    auto srcPointer = bitPattern.begin();
+    for (auto glyphNumber = 0u; glyphNumber < _glyphCount; glyphNumber++)
+    {
+        // We start by calculating the position in the bitmap where the glyph
+        // needs to be stored.
+        const auto xOffset = _xOffsetForGlyph<size_t>(glyphNumber);
+        const auto yOffset = _yOffsetForGlyph<size_t>(glyphNumber);
+        auto dstPointer = bitmapScanline(yOffset) + xOffset;
+        for (auto y = 0; y < sourceSize.height; y++)
+        {
+            // Then for each scanline in the source, we need expand the bits
+            // into 32-bit BGRA values. For every bit that is set we write out
+            // an FFFFFFFF value, and it not set, we write out 00000000. In the
+            // end, though, all we care about is the red component, since we'll
+            // later remap these values with a color matrix.
+            auto srcBits = *(srcPointer++);
+            for (auto x = 0; x < sourceSize.width; x++)
+            {
+                const auto srcBitIsSet = (srcBits & 0x8000) != 0;
+                *(dstPointer++) = srcBitIsSet ? 0xFFFFFFFF : 0x00000000;
+                srcBits <<= 1;
+            }

```

Comment:
Oh and I have to confess that I don't quite understand this loop...
`srcBitIsSet` is true if `0b100000` is set in `srcBits` and on each iteration `srcBits` is shifted left. But this means that `srcBitIsSet` is always `false` for any `x > 6`. How does this work?

Reply:
Doh.